File: velolabs_repos/lattis_android/app/src/main/java/com/lattis/ellipse/presentation/ui/authentication/launch/LaunchActivityPresenter.java
package com.lattis.ellipse.presentation.ui.authentication.launch;

import com.lattis.ellipse.presentation.ui.base.activity.ActivityPresenter;

import java.util.concurrent.TimeUnit;

import javax.inject.Inject;

import io.reactivex.Observable;
import io.reactivex.disposables.Disposable;
import io.reactivex.functions.Consumer;
import io.reactivex.schedulers.Schedulers;


/**
 * Created by ssd3 on 10/3/17.
 */

public class LaunchActivityPresenter extends ActivityPresenter<LaunchActivityView> {

    private Disposable launchScreenTimerSubscription=null;
    private static final int MILLISECONDS_FOR_LAUNCH_SCREEN = 3000;

    @Inject
    public LaunchActivityPresenter() {
    }

    @Override
    protected void updateViewState() {
        super.updateViewState();
    }

    public synchronized void subscribeToLaunchScreenTimer(boolean active) {
        if (active) {
            if (launchScreenTimerSubscription == null) {
                launchScreenTimerSubscription = Observable.timer(MILLISECONDS_FOR_LAUNCH_SCREEN, TimeUnit.MILLISECONDS)
                        .subscribeOn(Schedulers.io())
                        .subscribe(new Consumer<Long>() {
                            public void accept(Long aLong) {
                                subscribeToLaunchScreenTimer(false);
                                if(view!=null)
                                    view.onLaunchScreenTimer();
                            }
                        }, new Consumer<Throwable>() {
                            @Override
                            public void accept(Throwable throwable) {
                                subscribeToLaunchScreenTimer(false);
                                subscribeToLaunchScreenTimer(true);
                            }
                        });
            }
        } else {
            if (launchScreenTimerSubscription != null) {
                launchScreenTimerSubscription.dispose();
                launchScreenTimerSubscription = null;
            }
        }
    }
}

File: velolabs_repos/operator-android/app/src/main/java/io/lattis/operator/presentation/launch/LaunchActivityPresenter.kt
package io.lattis.operator.presentation.authentication.launch

import io.lattis.domain.models.Fleet
import io.lattis.domain.usecase.fleet.GetUserSavedFleetUseCase
import io.lattis.operator.presentation.base.RxObserver
import io.lattis.operator.presentation.base.activity.ActivityPresenter
import io.reactivex.Observable
import io.reactivex.disposables.Disposable
import io.reactivex.schedulers.Schedulers
import java.util.concurrent.TimeUnit
import javax.inject.Inject

class LaunchActivityPresenter @Inject constructor(
    private val getUserSavedFleetUseCase: GetUserSavedFleetUseCase
) : ActivityPresenter<LaunchActivityView>() {
    private var launchScreenTimerSubscription: Disposable? = null
    var MILLISECONDS_FOR_LAUNCH_SCREEN = 3000
    var userSavedFleet:Fleet?=null



    @Synchronized
    fun subscribeToLaunchScreenTimer(active: Boolean) {
        if (active) {
            if (launchScreenTimerSubscription == null) {
                launchScreenTimerSubscription = Observable.timer(MILLISECONDS_FOR_LAUNCH_SCREEN.toLong(), TimeUnit.MILLISECONDS)
                    .subscribeOn(Schedulers.io())
                    .subscribe({
                        subscribeToLaunchScreenTimer(false)
                        getUserSavedFleet()
                    }) {
                        subscribeToLaunchScreenTimer(false)
                        subscribeToLaunchScreenTimer(true)
                    }
            }
        } else {
            if (launchScreenTimerSubscription != null) {
                launchScreenTimerSubscription?.dispose()
                launchScreenTimerSubscription = null
            }
        }
    }

    fun getUserSavedFleet() {
        subscriptions.add(
            getUserSavedFleetUseCase
                .execute(object : RxObserver<Fleet>(view, false) {
                    override fun onNext(newUserSavedFleet:Fleet) {
                        super.onNext(newUserSavedFleet)
                        userSavedFleet = newUserSavedFleet
                        view?.onUserSavedFleetSuccess()
                    }
                    override fun onError(e: Throwable) {
                        super.onError(e)
                        view?.onUserSavedFleetFailure()
                    }
                })
        )
    }



}